### PR TITLE
Add New Monokai Pro Theme

### PR DESCRIPTION
**Pull request by kadenbarlow**
I'm pretty new to the emacs community, but started using Doom recently (introduced to it by Doom Themes) and have loved it. I created these three themes for myself that I rotate through depending on my mood. Wondering if any of these are worth adding in? 

## doom-monokai-pro
doom-monokai-pro

edit: removed other two themes

### CODE REVIEW

**Comment by hlissner** [comment]
I'm not so sure about atom-dark -- it seems too similar to `doom-one` or `doom-vibrant` (both of which are based off of atom one dark). monokai and xcode would be welcome additions though.

**Comment by ema2159** [comment]
It would be better to create separate PRs though. Did you based the themes on their official palettes?

 If I'm not wrong, Monokai was requested quite sometime ago so it would be a great addition.

**Comment by kadenbarlow** [comment]
Sure, I'll move doom-xcode to a separate PR. 

I based Monokai based on their official palette, except for the different background color which I set darker than what they use.

I couldn't find an official palette for the xcode theme, so I used a tool to inspect xcode and pull the RGB values.

**Comment by ema2159** [comment]
@kadenbarlow I would think it's better to change Monokai's background to match the original one from the spec, just for the sake of keeping it as similar to the original as possible. I'll test it and use it to check all the faces work ok and that there's no contrast problems. Then if everything's ok I'll merge it! Thanks!

**Comment by kadenbarlow** [comment]
@ema2159 Sounds great. I updated the theme to match the Monokai Pro background and made a couple of other small tweaks so the theme more closely matches Monokai Pro. https://monokai.pro/vscode. 

Let me know if any faces need tweaking and I can make changes. Thanks!

**Comment by ema2159** [comment]
Quick question. Where did you grabbed the colors? Or how did you probed them?

**Comment by kadenbarlow** [comment]
On this site https://monokai.pro/.
MacOS has a digital color meter tool that I used to grab the values. For the actual background color I downloaded the theme in VS Code and inspected it there.

**Comment by minikN** [comment]
Hi.

What about all the different flavors of monokai-pro? Like Spectrum, Ristretto and so one. I'd love to see them in doom as well. I could create PR's for them if you'd like me to.

**Comment by ema2159** [comment]
Go ahead! We're reviewing this current one so it may take some time

**Comment by minikN** [comment]
> Go ahead! We're reviewing this current one so it may take some time

Hi @ema2159 . I've created a first "beta". It's not ready for a PR yet. However, seems to be I am too stupid to even load the theme properly. I've added 
```
(package! doom-monokai-pro-spectrum-theme
  :recipe (:host github :repo "minikN/doom-emacs-monokai-pro-spectrum"))
```
into my `packages.el` and `(setq doom-theme 'doom-monokai-pro-theme)` into my `config.el`.

This is what I get: `condition-case: Error in a Doom startup hook: doom-init-theme-h, (error "Unable to find theme file for ‘doom-monokai-pro-theme’")`

After looking @ @kadenbarlow  theme. I suspect it's not quite accurate. As far as I know every Monokai Pro flavor/filter only provides 6 colors (Apart from black to white of course): `red`, `yellow`, `orange`, `green`, `purple`, `cyan`.

For example for the `spectrum` filter that would be:
#fc618d - red
#7bd88f - green
#fd9353 - orange
#fce566 - yellow
#948ae3 - purple
#5ad4e6 - cyan

As far as I know every other color is just a darkened/lightened variant of the above, like this one:
image

So I'm wondering where @kadenbarlow got his colors from especially `pink` and `red` as there should be only one of the two.

Cheers.

File: themes/doom-monokai-pro-theme.el [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+   (doc-comments   (doom-lighten grey 0.14))
+   (constants      violet)
+   (functions      green)
+   (keywords       pink)
+   (methods        green)
+   (operators      pink)
+   (type           light-blue)
+   (strings        yellow)
+   (variables      orange)
+   (numbers        yellow)
```
Comment (minikN) [inline themes/doom-monokai-pro-theme.el]: Numbers should be violet, not orange, see:

image

File: themes/doom-monokai-pro-theme.el [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+   (comments       grey)
+   (doc-comments   (doom-lighten grey 0.14))
+   (constants      violet)
+   (functions      green)
+   (keywords       pink)
+   (methods        green)
+   (operators      pink)
+   (type           light-blue)
+   (strings        yellow)
+   (variables      orange)
```
Comment (minikN) [inline themes/doom-monokai-pro-theme.el]: Variables should just be white. Function/method parameters are orange when declaring a function. But not variables itself.

Comment (ema2159) [reply themes/doom-monokai-pro-theme.el]: Sometimes it's a little bit more difficult given that themes sometimes vary between modes (languages) and Emacs doesn't provide a robust API for that matter. I'll find some time to review this properly a the weekend.

**Comment by kadenbarlow** [comment]
@minikN You'll notice that red is used for the vc-deleted face, not for any of the syntax highlighting. I think that red is a good indicator of deleted lines in version control and green is good for additions in version control. Also, I agree with @ema2159 It can be hard to get a consistent look across different modes. I tried to get a good balance of matching Monokai and getting a good contrast between colors in the different modes that I tried out.

File: themes/doom-monokai-pro-theme.el (hunk 2 above)
Comment (minikN) [reply themes/doom-monokai-pro-theme.el]: That's true. However, at least for JS and PHP, Monokai Pro doesn't vary in regards to variables.
I think the main problem is that major modes don't support the depth of customization some themes require. I only have experience with `php-mode` in regards to Monokai Pro. For example the concatenation operator (`.`) can't be fontified with what `php-mode` offers at the moment. I've actually submitted two PRs ([594](https://github.com/emacs-php/php-mode/pull/594), [606](https://github.com/emacs-php/php-mode/pull/606)). The latter is being reviewed right now. With those two PRs merged, flexibility with how php syntax can be fontified will be greatly improved.

However I don't know what the state of other major modes is.

**Comment by minikN** [comment]
> @minikN You'll notice that red is used for the vc-deleted face, not for any of the syntax highlighting. I think that red is a good indicator of deleted lines in version control and green is good for additions in version control. Also, I agree with @ema2159 It can be hard to get a consistent look across different modes. I tried to get a good balance of matching Monokai and getting a good contrast between colors in the different modes that I tried out.

Alright. That clarified it. Would you be able to give an example where Monokai Pro varies between different modes? At least for JS and PHP I haven't noticed that.

Regardless, I would still recommend sticking to the original as close as possible. Monokai Pro uses `#ff6188` and `#a9dc76` (both with an alpha of `0.19` and `0.26`) for version control (at least the VSCode implementation).

**Comment by minikN** [comment]
> > Go ahead! We're reviewing this current one so it may take some time
> 
> Hi @ema2159 . I've created a first "beta". It's not ready for a PR yet. However, seems to be I am too stupid to even load the theme properly. I've added
> 
> ```
> (package! doom-monokai-pro-spectrum-theme
>   :recipe (:host github :repo "minikN/doom-emacs-monokai-pro-spectrum"))
> ```
> 
> into my `packages.el` and `(setq doom-theme 'doom-monokai-pro-theme)` into my `config.el`.
> 
> This is what I get: `condition-case: Error in a Doom startup hook: doom-init-theme-h, (error "Unable to find theme file for ‘doom-monokai-pro-theme’")`

@ema2159 Would you be able to help with that?

File: themes/doom-monokai-pro-theme.el (hunk 1 above)
Comment (ema2159) [reply themes/doom-monokai-pro-theme.el]: Indeed

**Comment by ema2159** [comment]
@minikN of course, I'll just need sometime. I'm a little busy with my new job but as soon as I can I'll lend you a hand.

**Comment by hlissner** [comment]
@minikN wrt

```elisp
(package! doom-monokai-pro-spectrum-theme
  :recipe (:host github :repo "minikN/doom-emacs-monokai-pro-spectrum"))
```

There are two problems with the theme's definition:

1. [Here](https://github.com/minikN/doom-emacs-monokai-pro-spectrum/blob/master/doom-monokai-pro-spectrum-theme.el#L30), you've named the theme `monokai-pro-spectrum`,
2. But [here](https://github.com/minikN/doom-emacs-monokai-pro-spectrum/blob/master/doom-monokai-pro-spectrum-theme.el#L168), you've announced it as `doom-monokai-pro-spectrum`

Try removing the `provide` line entirely (`def-doom-theme` provides the theme for you), and replace:

```diff
-(def-doom-theme monokai-pro-spectrum
+(def-doom-theme doom-monokai-pro-spectrum
```

**Comment by hlissner** [comment]
@minikN Oh, I forgot to mention a third issue: you need to add the theme's directory to `custom-theme-load-path` somewhere in the file. e.g.

```elisp
;;;###autoload
(add-to-list 'custom-theme-load-path (file-name-directory load-file-name))
```

**Comment by minikN** [comment]
> @minikN wrt
> 
> ```emacs-lisp
> (package! doom-monokai-pro-spectrum-theme
>   :recipe (:host github :repo "minikN/doom-emacs-monokai-pro-spectrum"))
> ```
> 
> There are two problems with the theme's definition:
> 
> 1. [Here](https://github.com/minikN/doom-emacs-monokai-pro-spectrum/blob/master/doom-monokai-pro-spectrum-theme.el#L30), you've named the theme `monokai-pro-spectrum`,
> 2. But [here](https://github.com/minikN/doom-emacs-monokai-pro-spectrum/blob/master/doom-monokai-pro-spectrum-theme.el#L168), you've announced it as `doom-monokai-pro-spectrum`
> 
> Try removing the `provide` line entirely (`def-doom-theme` provides the theme for you), and replace:
> 
> ```diff
> -(def-doom-theme monokai-pro-spectrum
> +(def-doom-theme doom-monokai-pro-spectrum
> ```

Thank you @hlissner. I will try this tonight.

Is there any guideline as to how specific a theme should be? I'm talking about mode-specific faces.
Monokai Pro is pretty specific in how it highlights syntax for various modes.  As I [said](https://github.com/hlissner/emacs-doom-themes/pull/413#discussion_r388144344) some modes, in my case PHP mode didn't even support all the faces needed to implement Monokai Pro properly until I created a PR for it.

I have a [local](https://github.com/minikN/emacs-monokai-pro-theme/blob/master/monokai-pro-theme.el) repo for Monokai Pro in which I've added several new faces to PHP mode based on my PR:

´´´
;;;; php-mode
   (php-block-delimiter				 :foreground fg-3)
   (php-block-statement				 :foreground red)
   (php-class-declaration-spec 			 :foreground red)
   (php-class-modifiers 			 :foreground red)
   (php-constant-assign 			 :foreground red)
   (php-constant-keyword 			 :foreground purple)
   (php-flow-control-statement			 :foreground red)
   (php-import-declaration 			 :foreground red)
   (php-include-statement 			 :foreground green)
   (php-method-access				 :foreground red :slant italic)
   (php-method-modifiers			 :foreground red :slant italic)
   (php-method-static				 :foreground red :slant italic)
   (php-namespace-declaration			 :foreground red)
   (php-number					 :foreground purple)
   (php-php-tag 				 :foreground orange)
   (php-print-statement				 :foreground green)
   (php-property-access 			 :foreground red :slant italic)
   (php-property-const				 :foreground red :slant italic)
   (php-property-static				 :foreground red :slant italic)
   (php-string-quote				 :foreground fg-3)
   (php-type-operator				 :foreground red)
   (php-function-keyword 			 :foreground blue :slant italic)

   (php-function-name				 :foreground green)
   (php-function-call				 :foreground green)
   (php-string 					 :foreground yellow)
   (php-keyword 				 :foreground blue)
   (php-builtin 				 :foreground purple)
   (php-method-call 				 :foreground green)
   (php-static-method-call 			 :foreground green)
   (php-variable-name 				 :foreground fg)
   (php-property-name 				 :foreground fg)
   (php-variable-sigil 				 :foreground fg-2)
   (php-operator                                 :foreground red)
   (php-paamayim-nekudotayim 			 :foreground red)
   (php-type 					 :foreground blue :slant italic)
   (php-class 					 :foreground red)
   (php-constant 				 :foreground purple)
   (php-constant-assign 			 :foreground blue)
   (php-magical-constant 			 :foreground purple)
   (php-$this 					 :foreground fg-2 :slant italic)
   (php-$this-sigil				 :foreground fg-2 :slant italic)
   (php-errorcontrol-op 			 :foreground red)
   (php-doc-annotation-tag 			 :foreground blue)
   (php-doc-variable-sigil 			 :foreground fg-4)
   (php-doc-$this 				 :foreground fg-4)
   (php-doc-$this-sigil 			 :foreground fg-4)
   (php-doc-class-name 				 :foreground fg-4)
´´´

So my question is whether or not there is a threshold as to how fine grained a theme should be.

Thanks.

**Comment by hlissner** [comment]
> So my question is whether or not there is a threshold as to how fine grained a theme should be.

It can be as fine grained as you like, so long as you avoid properties that could throw errors, e.g. `:font` or `:family` when the specified font is missing on the user's system.

That said, I wasn't aware of php-mode's faces, perhaps they should be added to doom's base theme as well, so all themes can provide rudimentary support for the mode.

**Comment by minikN** [comment]
> > So my question is whether or not there is a threshold as to how fine grained a theme should be.
> 
> It can be as fine grained as you like, so long as you avoid properties that could throw errors, e.g. `:font` or `:family` when the specified font is missing on the user's system.
> 
> That said, I wasn't aware of php-mode's faces, perhaps they should be added to doom's base theme as well, so all themes can provide rudimentary support for the mode.

Thank you @hlissner . Well as I said, [https://github.com/emacs-php/php-mode/pull/594](594) is merged already, however [https://github.com/emacs-php/php-mode/pull/606](606) is still in review. 

Unfortunately, I'm still stuck with loading the theme. I applied the [changes](https://github.com/minikN/doom-emacs-monokai-pro-spectrum/commit/357a54aec4a7d215344b9cb2fa48f36a50358eb6) you mentioned.

If I now repace `(setq doom-theme 'doom-one)` with `(setq doom-theme 'doom-monokai-pro-spectrum)` in `config.el`I get this: `condition-case: Error in a Doom startup hook: doom-init-theme-h (wrong-type-argument stringp nil)`.

**Comment by hlissner** [comment]
@minikN That likely means `load-file-name` in `(file-name-directory load-file-name)` is nil. Change it to

```lisp
;;;###autoload
(when (and (boundp 'custom-theme-load-path) load-file-name)
  (add-to-list 'custom-theme-load-path
               (file-name-as-directory (file-name-directory load-file-name))))
```

**Comment by minikN** [comment]
> @minikN That likely means `load-file-name` in `(file-name-directory load-file-name)` is nil. Change it to
> 
> ```lisp
> ;;;###autoload
> (when (and (boundp 'custom-theme-load-path) load-file-name)
>   (add-to-list 'custom-theme-load-path
>                (file-name-as-directory (file-name-directory load-file-name))))
> ```

This worked. Thanks!

**Comment by ema2159** [comment]
I improved the palette and theming overall. It should look almost the same as VS Code's.

**Comment by kadenbarlow** [comment]
@ema2159 thanks for making those changes. Sorry been super busy at work the past week or so.

**Comment by ema2159** [comment]
@kadenbarlow don't worry. Thanks for contributing!

**Pull request merged**